clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1487,6 +1488,18 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
                    std::back_inserter(ArgTys),
                    [](const ParmVarDecl *PVD) { return PVD->getType(); });
 
+    // TODO: enable template instantiation tree for this diagnostic
```

Comment:
Its a bit of a shame it is so easy to implement here :)  

The 'correct' way of getting the template instantiation tree is to make this check done by a 'new' handler that is run at 'check' time (or perhaps, a new mode of this one that just 'counts' and does nothing else).

Putting this diagnostic here means that accomplishing this 'TODO' requires re-implementing this patch.

Reply:
The problem that all handlers that run at 'check' time are "FieldCheckers". So, If I implement this diagnostic in FieldChecker, I will have to duplicate visiting/handling/decomposing logic of the DeclCreator and keep it always aligned (same problem with the new handler). Another variant is to move "DeclCreator" routine to the "check" time, but I'm not sure how it can be done, because it is connected with "BodyCreator" which needs body of "sycl_kernel" function that doesn't exist at "check" time.

Reply:
The idea is that you could either create a new Field Checker (like UnionChecker) to do this, and would just keep count.

The second part of the idea is to have the DeclCreator have 2 'modes', one where it actually creates the decl/parameters, and the other where it just 'counts' instead of doing Params.push_back.  You'd then add one configured for 'count' mode to the 'Check' section, and the other in its current location.

Reply:
Okay, adding second mode for DeclCreator seems better to me than a new FieldChecker, because it will save us from forgotten handlers if we for example add new types of kernel arguments, I'll give it a try. Thanks.

Reply:
Is it intention that all "FieldHandlers" have all "handleType" functions are marked with final? I would rather inherit from DeclCreator instead of doing "templateize?" (I guess it is template specialization, right?), but it seems I will need to modify `handleSyclAccessorType` so it doesn't do `handleAccessorPropertyList` and any stuff under it and I can't do it because of final.

Reply:
They are marked 'final', as they are passed by reference and we get to avoid the indirection if they are final.
(That is, FieldHandler->DeclCreator->SomethingElse.... If a function is "final", a DeclCreator& doesn't have to use the V-table).

Would it be possible to extract a common-base type for the two?  The templatized thought was that you'd have template-specializations of any functions that 'changed', based on some class-level template parameter.

Reply:
My attempts to add second mode for DeclCreator ended with some weird ugly stuff, so I've added a new FieldHandler.